<a href="https://colab.research.google.com/github/aryan-0018/FlyRank-ML-W1/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/aryan-0018/FlyRank-ML-W1/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

### My Lane

I chose **Lane 2: Refresh / Content Opportunity Scoring**.

I selected this lane because it focuses on identifying which content pages should be reviewed first using observable search, content, and engagement signals. It builds naturally on the starter notebooks, where I explored declining pages and compared a hand-written prioritization rule with a machine learning model.

The unit of analysis for my project will be an individual content page. My goal is to develop an evidence-based and explainable way to prioritize pages for human review rather than automatically deciding that a page must be changed.

In [6]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

### Research Question

**Question:** Which content pages should be prioritized for human review based on observable search, content, and engagement signals?

**Decision:** The work aims to improve the decision of which pages should be reviewed first when review capacity is limited.

**Output:** The intended output is a ranked list of content pages, supported by scores and interpretable reason codes that explain why each page was prioritized.

**Action:** A content reviewer can inspect the highest-ranked pages and decide whether each page should be refreshed, expanded, protected, monitored, or considered for another appropriate content action.

**Cost of a wrong recommendation:** A false positive could waste limited review time by prioritizing a page that does not require attention. A false negative could cause a potentially important page to be overlooked. Because of these risks, the ranking should support human decision-making rather than automatically trigger content changes.

**Why data or machine learning can help:** Page performance depends on multiple signals such as visibility, freshness, position, CTR, and engagement. A transparent manual rule provides a useful baseline, while data analysis and machine learning can test whether combining these observable signals produces a more useful prioritization. The goal is not simply to train a model, but to determine whether an evidence-based ranking can improve a real review decision and remain understandable to the person using it.

In [7]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [8]:
import os
import subprocess
import pandas as pd

REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if not os.path.isdir(REPO_DIR):
    subprocess.run(
        ["git", "clone", "--depth", "1", REPO_URL, REPO_DIR],
        check=True
    )

data_path = os.path.join(
    REPO_DIR,
    "data",
    "raw",
    "content_refresh_anonymized.csv"
)

df = pd.read_csv(data_path)

print("Dataset loaded successfully.")
print("Rows:", len(df))
print("Columns:", len(df.columns))

declining_pages = (
    df["trend_direction"]
    .astype(str)
    .str.lower()
    .eq("down")
    .sum()
)

stale_visible_pages = (
    (df["days_since_last_update"] >= 180) &
    (df["impressions_90d"] >= 500)
).sum()

print("Total pages:", len(df))
print("Pages with a declining trend:", declining_pages)
print("Stale and visible pages:", stale_visible_pages)


Dataset loaded successfully.
Rows: 30000
Columns: 44
Total pages: 30000
Pages with a declining trend: 16262
Stale and visible pages: 17


### Quick Look at the Data

The starter dataset contains **30,000 pages**. Among them, **16,262 pages (approximately 54.2%)** have an observed `trend_direction` of "down".

Using a simple "stale and visible" rule — pages that have not been updated for at least 180 days and have at least 500 impressions over the 90-day measurement window — I found only **17 pages** that meet both conditions.

These observations make the prioritization problem interesting. While a large share of pages show a declining trend, only a small number satisfy this particular hand-written rule. This suggests that relying on a single strict rule may overlook other potentially relevant review candidates. Over the next several weeks, I want to investigate whether combining multiple observable signals can produce a more useful and explainable ranking of pages for human review.

These findings are exploratory and directional. A declining trend does not necessarily mean that a page should be refreshed, and meeting or not meeting the "stale and visible" rule does not establish that an intervention would improve future performance.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

### What I Can and Cannot Claim

This project can identify **observed associations and directional patterns** in the provided search, content, and engagement data. It can evaluate whether these observable signals help prioritize pages for review and can provide decision-support through an explainable ranked list.

The project cannot claim that any signal is a causal Google ranking factor, nor can it claim to predict Google's ranking algorithm. A page being ranked highly for review does not prove that it needs to be changed, and a declining trend does not prove that refreshing the page will cause it to recover.

Any recommendations produced by the project should therefore be treated as **decision-support for human review**, not as automated conclusions or guarantees of future performance.

In [9]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.